File: backend/src/main/java/com/tunisia/financial/service/TransactionService.java
package com.tunisia.financial.service;

import com.tunisia.financial.dto.transaction.TransactionRequest;
import com.tunisia.financial.dto.transaction.TransactionResponse;
import com.tunisia.financial.dto.transaction.TransactionStatistics;
import com.tunisia.financial.entity.Transaction;
import com.tunisia.financial.entity.User;
import com.tunisia.financial.enumerations.TransactionStatus;
import com.tunisia.financial.enumerations.TransactionType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

/**
 * Service interface for transaction management
 */
public interface TransactionService {
    
    /**
     * Create a new transaction
     */
    TransactionResponse createTransaction(TransactionRequest request, User user);
    
    /**
     * Get transaction by ID
     */
    TransactionResponse getTransactionById(Long id, User user);
    
    /**
     * Get all transactions with pagination
     */
    Page<TransactionResponse> getAllTransactions(Pageable pageable);
    
    /**
     * Get transactions for a specific user
     */
    Page<TransactionResponse> getTransactionsByUserId(UUID userId, Pageable pageable);
    
    /**
     * Get transactions by status
     */
    Page<TransactionResponse> getTransactionsByStatus(TransactionStatus status, Pageable pageable);
    
    /**
     * Get transactions by type
     */
    Page<TransactionResponse> getTransactionsByType(TransactionType type, Pageable pageable);
    
    /**
     * Get transactions for user by status
     */
    Page<TransactionResponse> getUserTransactionsByStatus(UUID userId, TransactionStatus status, Pageable pageable);
    
    /**
     * Get transactions for user by type
     */
    Page<TransactionResponse> getUserTransactionsByType(UUID userId, TransactionType type, Pageable pageable);
    
    /**
     * Get transaction by reference number
     */
    TransactionResponse getTransactionByReferenceNumber(String referenceNumber);
    
    /**
     * Get transactions within date range
     */
    Page<TransactionResponse> getTransactionsByDateRange(Instant startDate, Instant endDate, Pageable pageable);
    
    /**
     * Get user transactions within date range
     */
    Page<TransactionResponse> getUserTransactionsByDateRange(UUID userId, Instant startDate, Instant endDate, Pageable pageable);
    
    /**
     * Update transaction status
     */
    TransactionResponse updateTransactionStatus(Long id, TransactionStatus status, User user);
    
    /**
     * Cancel a pending transaction
     */
    TransactionResponse cancelTransaction(Long id, User user);
    
    /**
     * Get transaction statistics
     */
    TransactionStatistics getTransactionStatistics();
    
    /**
     * Get user transaction statistics
     */
    TransactionStatistics getUserTransactionStatistics(UUID userId);
    
    /**
     * Convert Transaction entity to TransactionResponse DTO
     */
    TransactionResponse convertToResponse(Transaction transaction);
}

File: frontend/src/services/transactionService.ts
import api from './api';
import {
  TransactionRequest,
  TransactionResponse,
  TransactionStatistics,
  TransactionStatus
} from '../types/transaction.types';
import { PaginatedResponse, SearchParams } from '../types/common.types';

const API_BASE = '/transactions';

export const transactionService = {
  // Create transaction
  createTransaction: async (data: TransactionRequest): Promise<TransactionResponse> => {
    const response = await api.post(API_BASE, data);
    return response.data;
  },

  // Get transaction by ID
  getTransactionById: async (id: number): Promise<TransactionResponse> => {
    const response = await api.get(`${API_BASE}/${id}`);
    return response.data;
  },

  // Get all transactions (Admin/Auditor/Analyst)
  getAllTransactions: async (params?: SearchParams): Promise<PaginatedResponse<TransactionResponse>> => {
    const response = await api.get(API_BASE, { params });
    return response.data;
  },

  // Get transactions by user ID
  getTransactionsByUserId: async (userId: string, params?: SearchParams): Promise<PaginatedResponse<TransactionResponse>> => {
    const response = await api.get(`${API_BASE}/user/${userId}`, { params });
    return response.data;
  },

  // Search transactions - Note: Backend doesn't have dedicated search endpoint, using base with params
  searchTransactions: async (query: string, params?: SearchParams): Promise<PaginatedResponse<TransactionResponse>> => {
    const response = await api.get(API_BASE, {
      params: { search: query, ...params }
    });
    return response.data;
  },

  // Get transactions by status
  getTransactionsByStatus: async (status: TransactionStatus, params?: SearchParams): Promise<PaginatedResponse<TransactionResponse>> => {
    const response = await api.get(`${API_BASE}/status/${status}`, { params });
    return response.data;
  },

  // Create batch transactions - Note: Backend doesn't have batch endpoint yet
  createBatchTransactions: async (data: TransactionRequest[]): Promise<TransactionResponse[]> => {
    const results = await Promise.all(
      data.map(transaction => api.post(API_BASE, transaction).then(res => res.data))
    );
    return results;
  },

  // Get transaction statistics (Admin/Analyst)
  getTransactionStatistics: async (): Promise<TransactionStatistics> => {
    const response = await api.get(`${API_BASE}/statistics`);
    return response.data;
  },

  // Update transaction status (Admin only) - Backend uses PUT with query param, not PATCH with body
  updateTransactionStatus: async (id: number, status: TransactionStatus): Promise<TransactionResponse> => {
    const response = await api.put(`${API_BASE}/${id}/status`, null, { params: { status } });
    return response.data;
  }
};
